### ISSUE TITLE
Improve method that determines whether a vehicle is supported properly.

### ISSUE THREAD

**Issue by sparr** [CleverRaven/Cataclysm-DDA#11687]
If the center of mass of a vehicle is outside the convex hull of the wheels then it will tip over[1]. We could check for this in valid_wheel_config and grind the vehicle to a halt when it happens, or we could apply a (significant) friction penalty, or some other result?

[1] unless it's turning so that centripetal force keeps it upright, which would be even more fun
##

**Comment by sparr** [comment]
An even more creative solution might be required for boats that tip over :)

**Comment by BevapDin** [comment]
Something similar is already done in `vehicle::valid_wheel_config`, but it's not that elaborated.

**Comment by sparr** [comment]
@BevapDin not sure how I missed that part. You're right that it's not as elaborate. Also, it re-implements the center of mass function?!?

The most common use case is a 4-wheel car that loses a wheel. The current system says that's still valid. In reality, that's only viable if the missing wheel is on the lighter half of the car.

I'm not sure if the extra complexity is worthwhile, so I'll wait for someone else to chime in here before I decide to implement this.

**Comment by BevapDin** [comment]
> Also, it re-implements the center of mass function?!?>

Yeah, I was wondering about that myself, it can probably be changed to use the appropriate function. The current code does not include the weight of the items, only the weight of the vehicle parts is used there.

> I'm not sure if the extra complexity is worthwhile, so I'll wait for someone else to chime in here before I decide to implement this.

Well, if you do it, cache the results as the the calculation is probably expensive. Consider a 20x20 vehicle with several dozen wheels.

**Comment by sparr** [comment]
The tipping over was more of a depiction of the real world result. In-game it might just be the same as not having enough wheels, grinding to a halt and not being able to drive any more. I'm not sure what happens in-game if you lose a necessary wheel while in motion, but that same thing would apply here.

I agree with caching the results.

**Comment by KA101** [comment]
> I'm not sure what happens in-game if you lose a necessary wheel while in motion, but that same thing would apply here.

Just tested: lose one wheel on a Car, no problem.  Lose wheels diagonally opposite each other (driver fore and passenger aft), no problem, surprisingly.  Lose driver aft as well, insufficient wheels, the Car ground into the grass within a turn and bounced back, raising dirt piles fore and aft.

So this is kinda already implemented.

**Comment by sparr** [comment]
@KA101 yeah, if you lost two wheels on the same side (left, right, front, back) I think it would fail now. The unrealism right now is that it uses the bounding box of the wheels instead of the convex hull.

**Comment by KA101** [comment]
You asked, I answered. :-|

**Comment by sparr** [comment]
if #11691 lands, it includes a //TODO mentioning the convex hull would be better than the bounding box of the wheels.

**Comment by kevingranade** [comment]
What would the benefit of using a convex hull be?

**Comment by sparr** [comment]
With the bounding box, you can make a square/rectangular vehicle have wheels at just two opposite corners and it will be "balanced" (have a valid_wheel_config) no matter where the CoM is. Using a convex hull of the two wheels, the CoM would have be on the diagonal connecting the two wheels. If it was in one of the other corners, the vehicle would (realistically) tip over and not be driveable/rollable.

**Comment by kevingranade** [comment]
Got it, yea that's an improvement.

**Comment by kevingranade** [comment]
Hypothetically an improvement but not enough to keep an issue about it around indefinitely.
